### PR TITLE
Support for Claude Code Switcher

### PR DESCRIPTION
**Pull request by felipeadeildo**
Adds native support for **[CCS (Claude Code Switcher)](https://github.com/kaitranntt/ccs)**, a tool that lets users manage multiple Claude accounts as named profiles (e.g., `work`, `home`, `personal`).

Without this PR, the widget only tracks the default `~/.claude/projects` directory. With CCS, users have additional profile directories under `~/.ccs/instances/<profile>/projects/`, and their usage was invisible to the widget.

<table>
  <tr>
    <td>image 1</td>
    <td>image 2</td>
  </tr>
</table>

### `get-claude-usage`
- Discovers CCS profile instances from `~/.ccs/instances/`
- Runs token counting and usage/subscription fetching **in parallel** for each profile
- Aggregates totals across all profiles into the existing top-level output fields
- Emits new `PROFILE_*` output fields with per-profile breakdowns:
  - `PROFILES`: comma-separated list of active profile names
  - `PROFILE_WEEK_TOKENS`, `PROFILE_MONTH_TOKENS`: per-profile token counts
  - `PROFILE_TODAY_COST`, `PROFILE_WEEK_COST`, `PROFILE_MONTH_COST`: per-profile costs
  - `PROFILE_DAILY`, `PROFILE_DAILY_COSTS`: per-profile 7-day arrays (pipe-separated)
  - `PROFILE_WEEK_MODELS`: per-profile model breakdown
  - `PROFILE_SUBSCRIPTION`, `PROFILE_TIER`, `PROFILE_FIVE_HOUR_UTIL`, `PROFILE_SEVEN_DAY_UTIL`, etc.: per-profile rate-limit and subscription info read from each profile's own credentials file
- Sanitizes profile names (strips `,`, `|`, `:`) to prevent delimiter injection in output fields

> **⚠️ Minor breaking change:** `WEEK_MODELS` now uses `=` instead of `:` as the model/count separator (e.g. `claude-sonnet-4-5=1200`). This was necessary because model names themselves contain colons (e.g. `claude-opus-4-5-20251001`), making `:` ambiguous. External scripts parsing `WEEK_MODELS` will need to update their separator.

### `ClaudeCodeUsageWidget.qml`
- `profileListModel` populated from `PROFILES`; first entry is always `"all"`
- Profile selector UI: tab bar for ≤5 profiles, dropdown for >5
- `currentPd` reactive property: switches display between aggregate and per-profile data
- All display cards (`displayWeekTokens`, `displayMonthTokens`, `displayTodayCost`, `displayRateLimitTier`, `displaySubscriptionType`, etc.) derive from `currentPd`
- Chart shows per-profile daily **overlay bars** when a profile is selected
- `formatTier` / `formatSubscription` helpers handle new tier strings (`max`, `team`, `enterprise`)
- Tooltip "total" label is translated via `root.tr()`
- `PROFILE_EXTRA_USAGE` parsed as a proper boolean via `parseProfileBool` helper

### Tests
New coverage in `tests/test-get-claude-usage.sh`: profile discovery, `PROFILE_WEEK_TOKENS` format/values, `PROFILE_DAILY` pipe-separator, `PROFILE_WEEK_MODELS` `=` separator, plus upstream edge-case suite (week/month boundary, malformed JSONL, session deduplication, multi-project aggregation, cache parsing, cost math).

## How it works with CCS

CCS stores each profile under `~/.ccs/instances/<name>/`:
```
~/.ccs/instances/
  work/
    projects/          # scanned for JSONL usage logs
    .credentials.json  # used for rate-limit/subscription fetch
  personal/
    projects/
    .credentials.json
```

The script detects these directories automatically — no configuration needed. If CCS is not installed, only the default profile is shown and the widget behaves exactly as before. All existing output fields remain unchanged; the new `PROFILE_*` fields are purely additive.

### CODE REVIEW

**Comment by felipeadeildo** [comment]
Up, @titeya i wuold like to know what do you think about that pr!!

File: ClaudeCodeUsageWidget.qml [hunk 1]
```diff
@@ -117,97 +195,350 @@ PluginComponent {
...
+        // Prefer tier label if it adds info beyond subType
+        if (tierLabel && tierLabel !== subLabel)
+            return subLabel + " · " + tierLabel;
+        return subLabel || tierLabel;
+    }
+
+    // Helper: parse "name:value,name:value,..." into profileData[name][field]
+    // For numeric fields. Full object replacement ensures QML reactivity.
+    function parseProfileSimple(val, field, isFloat) {
+        var _pd = Object.assign({}, profileData);
```
Comment (titeya) [inline ClaudeCodeUsageWidget.qml:279]: **blocking:** `Object.assign({}, profileData)` is a shallow copy — nested profile objects (e.g. `profileData["work"]`) are shared by reference. When you do `_pd[name][field] = v`, you mutate the original sub-object too. QML reactivity works here by accident because you reassign `profileData = _pd`, but it's fragile.

Please also clone the sub-object:
```js
if (!_pd[name]) _pd[name] = {};
else _pd[name] = Object.assign({}, _pd[name]);
_pd[name][field] = v;
```
Same fix needed in `parseProfileString` (line 301) and the `PROFILE_DAILY` / `PROFILE_DAILY_COSTS` / `PROFILE_WEEK_MODELS` cases.

File: get-claude-usage [hunk 2]
```diff
@@ -258,9 +202,237 @@ if [ -d "$PROJECTS_DIR" ]; then
...
+PROF_EXTRA_USAGE_LIST=""
+
+for idx in "${!TMPFILES[@]}"; do
+    tmpf="${TMPFILES[$idx]}"
+    usage_tmpf="${USAGE_TMPFILES[$idx]}"
+    pname="${PROFILE_NAMES[$idx]}"
+    [ -f "$tmpf" ] || continue
+
+    # Read per-profile token stats
+    p_wt=$(grep "^WEEK_TOKENS=" "$tmpf" 2>/dev/null | cut -d= -f2 || echo 0)
```
Comment (titeya) [inline get-claude-usage]: **blocking:** Most lines here use `cut -d= -f2` which only returns the text between the first and second `=`. This is fine for numeric fields, but `DAILY_COSTS` (line 360) contains values like `0.00,0.00,...` which could be affected if the format ever changes. More importantly, line 359 correctly uses `-f2-` for `WEEK_MODELS` (which contains `=` in its values), but the inconsistency is error-prone.

Please use `cut -d= -f2-` consistently on all lines for safety.

File: get-claude-usage [hunk 3]
```diff
@@ -242,13 +187,12 @@ if [ -d "$PROJECTS_DIR" ]; then
             first = 1
             for (m in week_model) {
                 if (week_model[m] > 0) {
-                    models = models (first ? "" : ",") m ":" week_model[m]
+                    models = models (first ? "" : ",") m "=" week_model[m]
```
Comment (titeya) [inline get-claude-usage:192]: **suggestion:** This changes the `WEEK_MODELS` separator from `:` to `=`, which is a breaking change for anything parsing the old format. Since the script and QML are deployed together as a plugin this is acceptable, but please add a brief comment here explaining *why* `=` was chosen (model names like `claude-3-5-sonnet:thinking` could contain colons).

File: ClaudeCodeUsageWidget.qml [hunk 4]
```diff
@@ -337,17 +668,164 @@ PluginComponent {
...
+                    font.pixelSize: Theme.fontSizeSmall
+                    font.weight: Font.Medium
+                    color: Theme.surfaceText
+                }
+            }
+
+            MouseArea {
+                anchors.fill: parent
+                cursorShape: Qt.PointingHandCursor
+                onClicked: profileDropdownPopup.visible = !profileDropdownPopup.visible
```
Comment (titeya) [inline ClaudeCodeUsageWidget.qml:770]: **suggestion:** The dropdown toggles on click but never closes when clicking outside. Consider wrapping it with a full-size transparent `MouseArea` that closes the popup, or using a QML `Popup` type which handles this natively.

File: README.md [hunk 5]
```diff
@@ -14,6 +14,9 @@ A [DMS (Dank Material Shell)](https://github.com/AvengeMedia/DankMaterialShell)
   - Weekly activity bar chart (Monday–Sunday) with interactive hover tooltips (token count + cost)
   - Per-model token usage for the current calendar week with dynamic model family detection
   - All-time session and message statistics
+- **CCS profile breakdown** — automatically discovers [CCS](https://github.com/AvengeMedia/ccs) instances in `~/.ccs/instances/` and shows per-profile token/cost stats with a hybrid profile selector (tabs for up to 4 profiles, dropdown for more)
```
Comment (titeya) [inline README.md]: **nit:** This links to `github.com/AvengeMedia/ccs` but your PR description references `github.com/kaitranntt/ccs`. Which one is the correct repo?

**Review by titeya — CHANGES_REQUESTED** [review]
Thanks for this contribution — the CCS multi-profile support is a solid addition. The parallel profiling with trap cleanup is well done, and the backward compatibility (no CCS = no change) is exactly right.

I have a few requested changes below, categorized by severity.

### Blocking
- Shallow copy bug in `parseProfileSimple` / `parseProfileString` (see inline)
- Inconsistent `cut -d= -f2` truncating values in the aggregation loop (see inline)

### Suggestions
- Add a comment explaining the `WEEK_MODELS` separator change (`:` → `=`)
- Dropdown should close on outside click
- Fix CCS link in README

### Nit
- The formatting changes (~40% of the QML diff) make review harder. Would you consider splitting them into a separate commit?

File: ClaudeCodeUsageWidget.qml (hunk 1 above)
Comment (felipeadeildo) [reply ClaudeCodeUsageWidget.qml:279]: Fixed in db5c69f. Applied `Object.assign({}, _pd[name])` clone before mutation in both `parseProfileSimple` and `parseProfileString`, and also in the three inline blocks (`PROFILE_DAILY`, `PROFILE_DAILY_COSTS`, `PROFILE_WEEK_MODELS`).

File: get-claude-usage (hunk 2 above)
Comment (felipeadeildo) [reply get-claude-usage]: Fixed in db5c69f. Changed all `cut -d= -f2` to `cut -d= -f2-` consistently across the entire profile aggregation loop (both token stats and usage/subscription blocks).

File: get-claude-usage (hunk 3 above)
Comment (felipeadeildo) [reply get-claude-usage:192]: Done in db5c69f. Added a comment right above the assignment explaining that `=` was chosen because model names like `claude-3-5-sonnet:thinking` can contain colons.

File: ClaudeCodeUsageWidget.qml (hunk 4 above)
Comment (felipeadeildo) [reply ClaudeCodeUsageWidget.qml:770]: Fixed in db5c69f. Added a full-size transparent `MouseArea` (id: `profileDropdownOverlay`) anchored to `root` at z=99, visible only when the popup is open. Clicking anywhere outside the popup closes it.

File: README.md (hunk 5 above)
Comment (felipeadeildo) [reply README.md]: Fixed in db5c69f. The correct repo is `kaitranntt/ccs` — updated both occurrences in README.md.

**Comment by felipeadeildo** [comment]
All review points addressed in db5c69f:

**Blocking**
- Shallow copy bug fixed in `parseProfileSimple`, `parseProfileString`, and the three inline blocks (`PROFILE_DAILY`, `PROFILE_DAILY_COSTS`, `PROFILE_WEEK_MODELS`) — sub-objects are now cloned with `Object.assign` before mutation
- `cut -d= -f2` → `cut -d= -f2-` applied consistently across the entire profile aggregation loop

**Suggestions**
- Added inline comment explaining the `=` separator choice for `WEEK_MODELS`
- Dropdown now closes on outside click via a full-size transparent `MouseArea` overlay anchored to `root` at z=99

**Nit**
- README CCS link corrected to `kaitranntt/ccs`

**Review by titeya — COMMENTED** [review]
Hey, a few things to fix before we can merge this.

### Blocking

**1. Profile name injection**
Profile names come from directory basenames (`~/.ccs/instances/*/`). A name containing `,`, `|`, or `:` would silently corrupt the pipe- and comma-delimited output fields (`PROFILE_DAILY`, `PROFILE_WEEK_TOKENS`, etc.). Please add sanitization or validation to reject/strip those characters.

**2. `WEEK_MODELS` is a breaking change — please update the description**
The PR says "All existing output fields are unchanged", but `WEEK_MODELS` format changed from `model:count` to `model=count`. Any external consumer parsing this field will break. The description should acknowledge this.

**3. `PROFILE_EXTRA_USAGE` should be stored as a boolean**
`parseProfileString` stores `"true"`/`"false"` as JS strings. A non-empty string is always truthy in QML, so `pd.extraUsageEnabled` will misbehave in boolean contexts. Please handle this with an explicit `=== "true"` conversion.

**4. "total" label is not translated**
In the chart tooltip, `t + " total"` is hardcoded in English. Please use `root.tr("total")` for consistency with the rest of the i18n setup.

### Non-blocking suggestions

- `wait "$pid" || true` silently swallows profile processing failures — a `stderr` warning would help debugging
- The Repeater model detection via identity comparison (`modelListData === modelRepeater.model`) is fragile — worth a comment at minimum
- Tests for profile names with special characters and for `PROFILE_FIVE_HOUR_RESET` ISO timestamp parsing would be a nice addition

**Comment by felipeadeildo** [comment]
All four blocking issues addressed in 2a8c8a0:

**1. Profile name injection**
`name=$(basename "$inst" | tr -d ',|:')` strips the delimiter characters from the directory basename before it ever reaches any output field. An empty result (basename was all forbidden chars) is also skipped.

**2. `WEEK_MODELS` breaking change**
The separator change from `:` to `=` was intentional and necessary: model names like `claude-opus-4-5-20251001` contain colons, so `:` was genuinely ambiguous as a delimiter. The PR description was imprecise -- I'll update it to acknowledge this as a deliberate, minor breaking change for external parsers.

**3. `PROFILE_EXTRA_USAGE` as boolean**
Rather than patching the inline block, I extracted a `parseProfileBool` helper that mirrors `parseProfileSimple` / `parseProfileString` -- same structure, same clone pattern, but stores `=== "true"` at the end. This also let me remove redundant inner `Object.assign` calls from the two existing helpers: the outer `Object.assign({}, profileData)` already isolates the copy, so cloning nested objects before writing was pure overhead.

`PROFILE_EXTRA_USAGE` now calls `parseProfileBool(val, "extraUsageEnabled")`; one line, consistent with every other `PROFILE_*` case in the switch.

**4. `"total"` not translated**
`t + " " + root.tr("total")` in the tooltip. No French translation needed since "total" is the same word.

Non-blocking suggestions noted, happy to add a `stderr` warning for `wait "$pid" || true` and a comment on the Repeater model comparison if you'd like those before merge.

**Review by titeya — APPROVED** [review]
LGTM after the final fixes (sub-object clone in the three `parseProfile*` helpers, `"total"` translation key, 1.2.0 bump). Thanks for the contribution!
